### PR TITLE
KAFKA-16696: Removed the in-memory implementation of RSM and RLMM

### PR DESCRIPTION
**Pull request by kamalcph**
The in-memory implementation of RSM and RLMM were written to write the unit/integration tests: https://github.com/apache/kafka/pull/10218

This is not used by any of the tests and superseded by the LocalTieredStorage framework which uses local-disk as secondary storage and topic as RLMM. Using the LocalTieredStorage framework is the preferred way to write the integration tests to capture any regression as it uses the internal topic as storage for RLMM which is the default implementation. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManagerTest.java
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManagerTest.java:52]: It seems to me `TopicBasedRemoteLogMetadataManagerWrapperWithHarness` can be removed also. We don't use the wrapper actually. This test can be modified by following style:

```java
    @Test
    public void testFetchSegments() throws Exception {
        try (TopicBasedRemoteLogMetadataManagerHarness remoteLogMetadataManagerHarness = new TopicBasedRemoteLogMetadataManagerHarness()) {
            RemoteLogMetadataManager remoteLogMetadataManager = remoteLogMetadataManagerHarness.remoteLogMetadataManager();
```

noted `TopicBasedRemoteLogMetadataManagerHarness` needs to implement `AutoClosable`, and `remoteLogMetadataManager` should be a public method.

**Review by chia7712 — COMMENTED** [review]
@kamalcph nice cleanup. one idea is left. PATL

**Review by showuon — APPROVED** [review]
LGTM! Thanks.

**Review by satishd — APPROVED** [review]
Thanks @kamalcph for the cleanup. Inmemory implementations were added fro UTs in the early days while RSM and RLMM interfaces were added. Those are no more needed as we can use `LocalTieredStorage` for the same. LGTM.

**Comment by chia7712** [comment]
QA is re-triggered, and I will merge it if QA does not show the related error.

File: storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManagerTest.java
Comment (chia7712) [reply storage/src/test/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManagerTest.java:52]: This can be addressed as follow-up I think :)

**Review by chia7712 — APPROVED** [review]

**Comment by clolov** [comment]
Can you actually not remove this because I was starting to use it in the GetOffsetShellToolTest? The problem I faced with using the topic-based RLMM is that it requires the bootstrap-servers to initialise correctly. However, said bootstrap-servers are not present at initialisation time (due to determining the port dynamically)

**Comment by kamalcph** [comment]
> Can you actually not remove this because I was starting to use it in the GetOffsetShellToolTest? The problem I faced with using the topic-based RLMM is that it requires the bootstrap-servers to initialise correctly. However, said bootstrap-servers are not present at initialisation time (due to determining the port dynamically)

We can get the port from the Kafkabroker using the `boundPort` method. If it does not work for you, Could you please open a draft PR? I'll take a look.

**Comment by chia7712** [comment]
> Can you actually not remove this because I was starting to use it in the GetOffsetShellToolTest? The problem I faced with using the topic-based RLMM is that it requires the bootstrap-servers to initialise correctly. However, said bootstrap-servers are not present at initialisation time (due to determining the port dynamically)

@clolov  Could you take a look at #15917 ?  I try to use another way to write tests for storage module. `ClusterInstace` get created and port is bound in testing phase. Maybe that can fix problem you described.

**Comment by clolov** [comment]
Okay, let's go forward with this so that you are not blocked on me, I will try to figure out a way forward and if not I will write again :)

**Comment by chia7712** [comment]
> Okay, let's go forward with this so that you are not blocked on me, I will try to figure out a way forward and if not I will write again :)

@clolov thanks! Please feel free to ping me if you have a draft PR. I'd like to make `ClusterInstance` works for more test cases.

**Pull request merged**
